homeassistant/components/prometheus/__init__.py

```diff
@@ -185,6 +189,44 @@ def handle_event(self, event):
         )
         last_updated_time_seconds.labels(**labels).set(state.last_updated.timestamp())
 
+    def handle_entity_registry_updated(self, event):
+        """Listen for deleted, disabled or renamed entities and remove them from the Prometheus Registry."""
+        if (action := event.data.get("action")) in (None, "create"):
+            return
+
+        entity_id = event.data.get("entity_id")
+        _LOGGER.debug("Handling entity update for %s", entity_id)
+
+        metrics_entity_id = None
+
+        if action == "remove":
+            metrics_entity_id = entity_id
+        elif action == "update":
+            changes = event.data.get("changes")
+            if "entity_id" in changes:
+                metrics_entity_id = changes["entity_id"]
+            elif "disabled_by" in changes or "name" in changes:
+                metrics_entity_id = entity_id
+
+        if metrics_entity_id:
+            self._remove_labelsets(metrics_entity_id)
+
+    def _remove_labelsets(self, entity_id):
+        """Remove labelsets matching the given entity id from all metrics."""
+        for _, metric in self._metrics.items():
+            for sample in metric.collect()[0].samples:
+                if sample.labels["entity"] == entity_id:
+                    _LOGGER.debug(
+                        "Removing labelset from %s for entity_id: %s",
+                        sample.name,
+                        entity_id,
+                    )
+                    try:
+                        metric.remove(*sample.labels.values())
+                    except KeyError:
+                        pass
+                    break
```

Comment:
I believe this will result in only the first labelset with a given entity label set being removed, while we need to remove all of them.

Reply:
Good catch! I didn't think of that. The `break` could be removed or the functions checks if the metric name contains `climate_action` or `humidifier_mode`.

1. If the `break` is removed the `remove` functions will be called multiple times but the `KeyError` will be catched (for example for Counter Metrics: `state_change` with child metrics `state_change_created` and `state_change_total`)

2. If the metric name is checked the function has to be updated every time a new metric with multiple labelsets for the same entity is added

I think option 1 would be better?

In both cases the `friendly_name` should be added as a parameter to `_remove_labelsets` since it no longer removes just the oldest labelset. `handle_state_changed` could be faster than `handle_entity_registry_updated` and mistakenly remove the newly generated metric. With the combination of `entity_id` and `friendly_name` it should be unique.

Reply:
I updated `_remove_labelsets` with a `friendly_name` parameter and removed the `break` command.
The tests randomly failed from time to time. When it fails `entity_registry_updated` event returns returns `changes=name=None`  instead of `changes=name="Outside Temperature"` . I assume its a timing issue.
Because of that I moved the `friendly_name` update detection to `handle_state_changed` which seems to work as expected, but doesn't look as clean since its now handled in two events.

## Log
````
DEBUG:homeassistant.core:Bus:Handling <Event entity_registry_updated[L]: action=create, entity_id=sensor.outside_temperature>
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=sensor.outside_temperature, old_state=None, new_state=<state sensor.outside_temperature=15.6; state_class=measurement, battery_level=12, unit_of_measurement=°C, device_class=temperature, friendly_name=Outside Temperature @ 2021-12-09T17:52:05.071434+00:00>>

DEBUG:homeassistant.core:Bus:Handling <Event entity_registry_updated[L]: action=update, entity_id=sensor.outside_temperature, changes=name=None>
DEBUG:homeassistant.core:Bus:Handling <Event state_changed[L]: entity_id=sensor.outside_temperature, old_state=<state sensor.outside_temperature=15.6; state_class=measurement, battery_level=12, unit_of_measurement=°C, device_class=temperature, friendly_name=Outside Temperature @ 2021-12-09T17:52:05.071434+00:00>, new_state=<state sensor.outside_temperature=15.6; state_class=measurement, battery_level=12, unit_of_measurement=°C, device_class=temperature, friendly_name=Inside Temperature @ 2021-12-09T17:52:05.071434+00:00>>
```